### ISSUE TITLE
Compatibility with planned node.js native ESM import resolution

### ISSUE THREAD

**Issue by coreyfarrell** [xojs/xo#386]
It appears that the node.js plan for native ESM is that file extensions [will be mandatory](https://nodejs.org/dist/latest-v12.x/docs/api/esm.html#esm_mandatory_file_extensions).  This directly conflicts with current `xo` rules `import/extensions` and `unicorn/import-index`.  I'm opening this ticket to start the discussion about how/when `xo` should deal with this.

It appears that [import/extensions](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md) only applies to `import` statements and not `require`.  If/when [proposal-pkg-exports](https://github.com/jkrems/proposal-pkg-exports) sees progress `import/extensions` may need to be updated.  [unicorn/import-index](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/import-index.md) applies to both `import` and `require`, might need an option so it can be applied to `require` only if that is desired.  Might also need/want a rule that forbids direct import of the package main source.

In one of my private projects that is targeting ESM I have the following xo.rules in my package.json:
```js
	"import/extensions": [
		"error",
		"always",
		{
			"ignorePackages": true
		}
	],
	"unicorn/import-index": "off"
```

Personally I'm not a fan of this but I feel like the decision has already been made.

**Comment by sindresorhus** [comment]
I think we can just disable `import/extensions` in the next version. It's not that useful anyway. No one uses file extension in CommonJS. As for `unicorn/import-index`, we could add an option of whether to apply to `import` or not, PR welcome for that.

**Comment by coreyfarrell** [comment]
I know this would be a rule reversal but I think `xo` should report an error if ES module extensions are omitted, obvious exception for the bare package name as shown in my xo.rules above.  My reason for this is that `import something from './somefolder'` is an error in the browser and native node.js ES modules.  This does not effect CJS (`import/extensions` does not process CJS).

I've posted a bug to eslint-plugin-unicorn.  If nobody else posts a PR for it I'll try taking a look when I get a chance (likely after I get nyc@15 released).

**Comment by fregante** [comment]
I had to disable all of these to use `--experimental-modules` with AVA
```json
      "import/extensions": 0,
      "import/no-unassigned-import": 0,
      "import/no-useless-path-segments": 0,
      "unicorn/import-index": 0
```

**Comment by fregante** [comment]
Ideally XO should automatically disable those rules on modules (i.e. `.mjs` files and `type: "module"` directories)

**Comment by jimmywarting** [comment]
if it is some hybrid node/browser module you should definitely use extension and no index import. Using "never" is bad practice and should be removed IMO

**Comment by fregante** [comment]
`require` has always worked well and will continue to work well **without extensions**

```json
{
	"import/extensions": "error",
	"import/no-useless-path-segments": "error",
	"unicorn/import-index": "error"
}
```

`import` initially inherited `require`’s behavior and is still often mapped to it, but going forward perhaps XO should enforce the opposite **on ES Modules**:


```json
{
	"import/extensions": "off",
	"import/no-useless-path-segments": "off",
	"unicorn/import-index": "off"
}
```

ES modules being:

- `.mjs`
- `type: "module"`
- `import`/`export` found in the file

This detection however might not be easy in XO’s context.

**Comment by sindresorhus** [comment]
Detection will be too hard and not worth it. I'm just going to make this good by default.

**Comment by sindresorhus** [comment]
https://github.com/xojs/xo/releases/tag/v0.37.0

**Comment by aureooms** [comment]
Hello! Is there a tool that can be used to automatically patch existing codebases by resolving node syntax and replace it with mandatory file extensions?

**Comment by sindresorhus** [comment]
Maybe https://github.com/wessberg/cjstoesm#cli-usage (I have not tried it)

**Comment by aureooms** [comment]
> Maybe https://github.com/wessberg/cjstoesm#cli-usage (I have not tried it)

I took a look at it and it seems it does not resolve old import syntax. I see you have added an auto-fix to `xo` in f4f6b396a3a16101b81f339cccb2856438a0063a so I guess we will be able to use `xo --fix` soon to replace old import syntax with the new one. Thank you!

**Comment by wessberg** [comment]
For the record, the default behavior of [`cjstoesm`](https://github.com/wessberg/cjstoesm) is now to add extensions to module specifiers, as well as to resolve directory-imports, so you should be able to use that tool to convert your project from CJS to ESM. This aligns the default behavior with the behavior across browsers and node.js.
